torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -119,38 +119,31 @@ using namespace torch::distributed::autograd;
 c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
     const std::shared_ptr<FutureMessage>& futureResponseMessage,
     bool hasValue) {
-  // Save and pass thread local state into the callback. The thread local state
-  // is a lightweight structure when the profiler is disabled.
-  at::ThreadLocalState tls_state;
   if (hasValue) {
     c10::intrusive_ptr<JitFuture> jitFuture =
         c10::make_intrusive<JitFuture>(PyObjectType::get());
     futureResponseMessage->addCallback(
-        [jitFuture, tls_state = std::move(tls_state)](
-            const FutureMessage& futureResponseMessage) {
-          at::ThreadLocalStateGuard g(tls_state);
-          if (futureResponseMessage.hasError()) {
-            jitFuture->setError(futureResponseMessage.error()->what());
+        at::wrapPropagateTLSState<void>([jitFuture, futureResponseMessage]() {
```

Comment:
Actually - are we sure that this case (and some others) don't lead to a memory leak?

The ones that take "const FutureMessage&" arg - I think these were actually mostly on purpose. When I was refactoring the futures months ago, my initial check-in caused a memory regression, e.g. in ctr_mobile_feed.  =(

And it took a bit of effort to figure out which was causing the regression. IIRC, sometimes the callbacks weren't being [ever] called. Without the ref cycle, things would be ok - they'd just vanish. But when there's a reference cycle like this, they could never be freed, and we'd leak memory pretty quickly.

My suggestion here would be maybe take an explicit weak reference for futureResponseMesssage. Assert that the result is non-null when you promote it to strong shared_ptr. Then everything should be ok with the wrapping.

But if you take a strong shared_ptr reference in these cases, there might be some unfortunate side effects.

Reply:
I see, the reason that I refactored out the `const FutureMessage& futureResponseMessage` argument types was because the `wrapPropagateTLSState` API only accepts functions which take no args, and I thought that the `addCallback` API here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/future.h#L135 was mostly a helper function that this code is equivalent to. It looks like it's not equivalent though, because in the case where the callback isn't called we would never deallocate `futureResponseMessage`? 

In that function, it looks like we make a copy of the `this` pointer?

Reply:
Right, I understand the motivation with the new function wrapper - keeping without arguments is good, but we just have to avoid capturing a shared_ptr in the lambda captures.

The helper `const FutureMessage&` felt slightly hacky (though is safe in its uses) - it's effectively just passing in an unsafe pointer to avoid the reference cycle. (The reference cycle causing `futureResponseMessage` never to be dealloced, and a slow leak)

You could pass in the unsafe pointer into the lambda capture, though it might be better-received than if you captured the weak reference instead:
```
  std:weak_ptr<FutureMessage> wp = futureResponseMessage;
  addCallback([wp]() {
    auto futureResponseMessage = wp.lock();
    ...
  });
```